Source/WebSocket/hcwebsocket.cpp

```diff
@@ -468,28 +485,52 @@ void HC_WEBSOCKET::CloseFunc(
 )
 {
     UNREFERENCED_PARAMETER(context);
+
+    // We release the provider's reference if we handle the close.
+    bool shouldDecRef { false };
     {
         std::lock_guard<std::recursive_mutex> lock{ websocket->m_mutex };
 
+        auto state = websocket->m_state;
         websocket->m_state = State::Disconnected;
```

Comment:
>websocket->m_state = State::Disconnected; [](start = 8, length = 41)

should this only be in the bottom 2 cases?

Reply:
I think we shouldn't be transitioning states if we get an unexpected provider callback

---
In reply to: [499870395](https://github.com/microsoft/libHttpClient/pull/545#discussion_r499870395) [](ancestors = 499870395)

Reply:
That was the existing behavior. Moving it would result in a breaking change.
[https://github.com/microsoft/libHttpClient/blob/master/Source/WebSocket/hcwebsocket.cpp#L474](https://github.com/microsoft/libHttpClient/blob/master/Source/WebSocket/hcwebsocket.cpp#L474)

---
In reply to: [499871609](https://github.com/microsoft/libHttpClient/pull/545#discussion_r499871609) [](ancestors = 499871609,499870395)